# Chapter 8: Data processing excercise


In this lesson we will be using a number of things we have learned in previous lessons. Try to go back to those lessons/notebooks to solve each part of the assignement.

P.S. you can also search the internet/cheat sheets to find the appropriate code for what you want to do. 

Before we start, let's load our favourite libraries.

In [120]:
import pandas as pd
import numpy as np
import  matplotlib.pyplot as plt
import seaborn as sns; sns.set()
import scipy.stats 
import sklearn.preprocessing

## 8.1 Data & pre-processing

The data we will be using for this assignment is the IMDb dataset. IMDb (Internet Movie Database) contains information on almost all movies ever made. All these movies receive ratings from the public and from critics.

There are 2 datasets, 'IMDb movies.csv' and 'IMDb ratings.csv'. The names are self-explanatory.

Before diving in, let's get familiar with our data. Like checking a recipe before cooking, we'll look at the structure and any potential issues.

### 8.1.1 Data Loading

Load both datasets correctly. Show the top 5 rows of both datasets.

Here, we use `pd.read_csv()` with `sep=","` to specify the comma separator (common for CSV files). This tells pandas how to split the data into columns, like using a knife to cut ingredients.

In [121]:
movies = pd.read_csv("IMDb movies.csv",sep=",")
ratings = pd.read_csv("IMDb ratings.csv",sep=",")
movies.head(10)

/var/folders/76/z1mjvk1s7k36fwnyfkcvx1q40000gn/T/ipykernel_1900/3524116668.py:1: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  movies = pd.read_csv("IMDb movies.csv",sep=",")


,imdb_title_id,title,original_title,year,date_published,genre,duration,country,language,director,...,actors,description,avg_vote,votes,budget,usa_gross_income,worlwide_gross_income,metascore,reviews_from_users,reviews_from_critics
0,tt0000009,Miss Jerry,Miss Jerry,1894,1894-10-09,Romance,45,USA,NaN,Alexander Black,...,"Blanche Bayliss, William Courtenay, Chauncey D...",The adventures of a female reporter in the 1890s.,5.9,154,NaN,NaN,NaN,NaN,1.0,2.0
1,tt0000574,The Story of the Kelly Gang,The Story of the Kelly Gang,1906,1906-12-26,"Biography, Crime, Drama",70,Australia,NaN,Charles Tait,...,"Elizabeth Tait, John Tait, Norman Campbell, Be...",True story of notorious Australian outlaw Ned ...,6.1,589,$ 2250,NaN,NaN,NaN,7.0,7.0
2,tt0001892,Den sorte drøm,Den sorte drøm,1911,1911-08-19,Drama,53,"Germany, Denmark",NaN,Urban Gad,...,"Asta Nielsen, Valdemar Psilander, Gunnar Helse...",Two men of high rank are both wooing the beaut...,5.8,188,NaN,NaN,NaN,NaN,5.0,2.0
3,tt0002101,Cleopatra,Cleopatra,1912,1912-11-13,"Drama, History",100,USA,English,Charles L. Gaskill,...,"Helen Gardner, Pearl Sindelar, Miss Fielding, ...",The fabled queen of Egypt's affair with Roman ...,5.2,446,$ 45000,NaN,NaN,NaN,25.0,3.0
4,tt0002130,L'Inferno,L'Inferno,1911,1911-03-06,"Adventure, Drama, Fantasy",68,Italy,Italian,"Francesco Bertolini, Adolfo Padovan",...,"Salvatore Papa, Arturo Pirovano, Giuseppe de L...",Loosely adapted from Dante's Divine Comedy and...,7.0,2237,NaN,NaN,NaN,NaN,31.0,14.0
5,tt0002199,"From the Manger to the Cross; or, Jesus of Naz...","From the Manger to the Cross; or, Jesus of Naz...",1912,1913,"Biography, Drama",60,USA,English,Sidney Olcott,...,"R. Henderson Bland, Percy Dyer, Gene Gauntier,...","An account of the life of Jesus Christ, based ...",5.7,484,NaN,NaN,NaN,NaN,13.0,5.0
6,tt0002423,Madame DuBarry,Madame DuBarry,1919,1919-11-26,"Biography, Drama, Romance",85,Germany,German,Ernst Lubitsch,...,"Pola Negri, Emil Jannings, Harry Liedtke, Edua...","The story of Madame DuBarry, the mistress of L...",6.8,753,NaN,NaN,NaN,NaN,12.0,9.0
7,tt0002445,Quo Vadis?,Quo Vadis?,1913,1913-03-01,"Drama, History",120,Italy,Italian,Enrico Guazzoni,...,"Amleto Novelli, Gustavo Serena, Carlo Cattaneo...","An epic Italian film ""Quo Vadis"" influenced ma...",6.2,273,ITL 45000,NaN,NaN,NaN,7.0,5.0
8,tt0002452,Independenta Romaniei,Independenta Romaniei,1912,1912-09-01,"History, War",120,Romania,NaN,"Aristide Demetriade, Grigore Brezeanu",...,"Aristide Demetriade, Constanta Demetriade, Con...",The movie depicts the Romanian War of Independ...,6.7,198,ROL 400000,NaN,NaN,NaN,4.0,1.0
9,tt0002461,Richard III,Richard III,1912,1912-10-15,Drama,55,"France, USA",English,"André Calmettes, James Keane",...,"Robert Gemp, Frederick Warde, Albert Gardner, ...",Richard of Gloucester uses manipulation and mu...,5.5,225,$ 30000,NaN,NaN,NaN,8.0,1.0


In [122]:
ratings.head(10)

,imdb_title_id,weighted_average_vote,total_votes,mean_vote,median_vote,votes_10,votes_9,votes_8,votes_7,votes_6,...,females_30age_avg_vote,females_30age_votes,females_45age_avg_vote,females_45age_votes,top1000_voters_rating,top1000_voters_votes,us_voters_rating,us_voters_votes,non_us_voters_rating,non_us_voters_votes
0,tt0000009,5.9,154,5.9,6.0,12,4,10,43,28,...,5.7,13.0,4.5,4.0,5.7,34.0,6.4,51.0,6.0,70.0
1,tt0000574,6.1,589,6.3,6.0,57,18,58,137,139,...,6.2,23.0,6.6,14.0,6.4,66.0,6.0,96.0,6.2,331.0
2,tt0001892,5.8,188,6.0,6.0,6,6,17,44,52,...,5.8,4.0,6.8,7.0,5.4,32.0,6.2,31.0,5.9,123.0
3,tt0002101,5.2,446,5.3,5.0,15,8,16,62,98,...,5.5,14.0,6.1,21.0,4.9,57.0,5.5,207.0,4.7,105.0
4,tt0002130,7.0,2237,6.9,7.0,210,225,436,641,344,...,7.3,82.0,7.4,77.0,6.9,139.0,7.0,488.0,7.0,1166.0
5,tt0002199,5.7,484,5.8,6.0,33,15,48,80,123,...,5.1,13.0,6.5,15.0,5.5,67.0,5.9,173.0,5.6,181.0
6,tt0002423,6.8,753,6.8,7.0,80,65,105,209,142,...,7.2,40.0,8.2,34.0,6.2,88.0,6.6,139.0,6.8,455.0
7,tt0002445,6.2,273,6.2,6.0,15,8,30,74,66,...,6.3,11.0,6.2,5.0,5.7,41.0,5.8,40.0,6.2,154.0
8,tt0002452,6.7,198,7.1,7.0,53,7,23,35,30,...,6.8,10.0,6.8,6.0,5.4,25.0,6.7,25.0,6.8,135.0
9,tt0002461,5.5,225,5.4,6.0,9,6,22,34,53,...,7.2,6.0,9.7,9.0,5.4,37.0,5.8,109.0,5.4,68.0


### 8.1.0.5 Initial Data Checks

Before selecting columns, let's check for missing values and data types. This is like inspecting ingredients for quality before using them in a recipe. Missing data can cause errors later, so we address it early.

Use `info()` and `isnull().sum()` to explore the datasets.

In [123]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85855 entries, 0 to 85854
Data columns (total 22 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   imdb_title_id          85855 non-null  object 
 1   title                  85855 non-null  object 
 2   original_title         85855 non-null  object 
 3   year                   85855 non-null  object 
 4   date_published         85855 non-null  object 
 5   genre                  85855 non-null  object 
 6   duration               85855 non-null  int64  
 7   country                85791 non-null  object 
 8   language               84954 non-null  object 
 9   director               85768 non-null  object 
 10  writer                 84283 non-null  object 
 11  production_company     81400 non-null  object 
 12  actors                 85786 non-null  object 
 13  description            83740 non-null  object 
 14  avg_vote               85855 non-null  float64
 15  vo

In [124]:
movies.isnull().sum()

imdb_title_id                0
title                        0
original_title               0
year                         0
date_published               0
genre                        0
duration                     0
country                     64
language                   901
director                    87
writer                    1572
production_company        4455
actors                      69
description               2115
avg_vote                     0
votes                        0
budget                   62145
usa_gross_income         70529
worlwide_gross_income    54839
metascore                72550
reviews_from_users        7597
reviews_from_critics     11797
dtype: int64

In [125]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85855 entries, 0 to 85854
Data columns (total 49 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   imdb_title_id              85855 non-null  object 
 1   weighted_average_vote      85855 non-null  float64
 2   total_votes                85855 non-null  int64  
 3   mean_vote                  85855 non-null  float64
 4   median_vote                85855 non-null  float64
 5   votes_10                   85855 non-null  int64  
 6   votes_9                    85855 non-null  int64  
 7   votes_8                    85855 non-null  int64  
 8   votes_7                    85855 non-null  int64  
 9   votes_6                    85855 non-null  int64  
 10  votes_5                    85855 non-null  int64  
 11  votes_4                    85855 non-null  int64  
 12  votes_3                    85855 non-null  int64  
 13  votes_2                    85855 non-null  int

In [126]:
ratings.isnull().sum()

imdb_title_id                    0
weighted_average_vote            0
total_votes                      0
mean_vote                        0
median_vote                      0
votes_10                         0
votes_9                          0
votes_8                          0
votes_7                          0
votes_6                          0
votes_5                          0
votes_4                          0
votes_3                          0
votes_2                          0
votes_1                          0
allgenders_0age_avg_vote     52496
allgenders_0age_votes        52496
allgenders_18age_avg_vote      706
allgenders_18age_votes         706
allgenders_30age_avg_vote       10
allgenders_30age_votes          10
allgenders_45age_avg_vote       80
allgenders_45age_votes          80
males_allages_avg_vote           1
males_allages_votes              1
males_0age_avg_vote          58444
males_0age_votes             58444
males_18age_avg_vote          1465
males_18age_votes   

### 8.1.2 Data selection

There are a lot of variables but we do not need all of them. Keep only the following variables:

- movie data: 'imdb_title_id', 'title', 'year', 'genre', 'duration', 'country', 'language', 'avg_vote', 'votes'
- rating data: 'imdb_title_id', 'total_votes', 'weighted_average_vote', 'mean_vote', 'us_voters_rating', 'non_us_voters_rating'

How many rows and columns do we have remaining for each dataset?

In [127]:
movie_data = movies[['imdb_title_id','title','year','genre','duration','country','language','avg_vote','votes']]
movie_data.head()

,imdb_title_id,title,year,genre,duration,country,language,avg_vote,votes
0,tt0000009,Miss Jerry,1894,Romance,45,USA,NaN,5.9,154
1,tt0000574,The Story of the Kelly Gang,1906,"Biography, Crime, Drama",70,Australia,NaN,6.1,589
2,tt0001892,Den sorte drøm,1911,Drama,53,"Germany, Denmark",NaN,5.8,188
3,tt0002101,Cleopatra,1912,"Drama, History",100,USA,English,5.2,446
4,tt0002130,L'Inferno,1911,"Adventure, Drama, Fantasy",68,Italy,Italian,7.0,2237


In [128]:
rating_data = ratings[['imdb_title_id','total_votes','weighted_average_vote','mean_vote','us_voters_rating','non_us_voters_rating']]
rating_data

,imdb_title_id,total_votes,weighted_average_vote,mean_vote,us_voters_rating,non_us_voters_rating
0,tt0000009,154,5.9,5.9,6.4,6.0
1,tt0000574,589,6.1,6.3,6.0,6.2
2,tt0001892,188,5.8,6.0,6.2,5.9
3,tt0002101,446,5.2,5.3,5.5,4.7
4,tt0002130,2237,7.0,6.9,7.0,7.0
...,...,...,...,...,...,...
85850,tt9908390,398,5.3,5.5,6.3,5.3
85851,tt9911196,724,7.7,7.9,6.8,7.7
85852,tt9911774,265,7.9,7.8,NaN,2.0
85853,tt9914286,194,6.4,9.4,1.7,5.8


**Checkpoint:** You should have 85,855 rows × 6 columns for ratings. If not, review your column selection or check for loading issues (see Module 2).

### 8.1.3 Data cleaning

There is a movie with a wrong year. This will cause problems when fixing the dtypes of the data. Like a mislabeled ingredient, it can spoil the whole dish.

- What is the movie ID and title of that movie?
- Fix the year!

*If you cannot figure this out, execute the next line of code and proceed for now. Do come back later to figure it out! **Note:** this line of code is not a solution for the problem, it simply avoids it.*

`films_merged = films_merged[films_merged['imdb_title_id'] !='tt8206668']`

In [129]:
movie_data['year']=pd.to_numeric(movie_data['year'],errors='coerce').ffill()
movie_data['year'] = movie_data['year'].astype(int)

/var/folders/76/z1mjvk1s7k36fwnyfkcvx1q40000gn/T/ipykernel_1900/1661823462.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movie_data['year']=pd.to_numeric(movie_data['year'],errors='coerce').ffill()
/var/folders/76/z1mjvk1s7k36fwnyfkcvx1q40000gn/T/ipykernel_1900/1661823462.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movie_data['year'] = movie_data['year'].astype(int)


In [130]:
movie_data[movie_data['imdb_title_id'] == 'tt8206668']

,imdb_title_id,title,year,genre,duration,country,language,avg_vote,votes
83917,tt8206668,Bad Education,2019,"Biography, Comedy, Crime",108,USA,English,7.1,23973


In [131]:
movie_data

,imdb_title_id,title,year,genre,duration,country,language,avg_vote,votes
0,tt0000009,Miss Jerry,1894,Romance,45,USA,NaN,5.9,154
1,tt0000574,The Story of the Kelly Gang,1906,"Biography, Crime, Drama",70,Australia,NaN,6.1,589
2,tt0001892,Den sorte drøm,1911,Drama,53,"Germany, Denmark",NaN,5.8,188
3,tt0002101,Cleopatra,1912,"Drama, History",100,USA,English,5.2,446
4,tt0002130,L'Inferno,1911,"Adventure, Drama, Fantasy",68,Italy,Italian,7.0,2237
...,...,...,...,...,...,...,...,...,...
85850,tt9908390,Le lion,2020,Comedy,95,"France, Belgium",French,5.3,398
85851,tt9911196,De Beentjes van Sint-Hildegard,2020,"Comedy, Drama",103,Netherlands,"German, Dutch",7.7,724
85852,tt9911774,Padmavyuhathile Abhimanyu,2019,Drama,130,India,Malayalam,7.9,265
85853,tt9914286,Sokagin Çocuklari,2019,"Drama, Family",98,Turkey,Turkish,6.4,194


### 8.1.4 Data merging 

Merge both datasets into 1 DataFrame. Combine the dataframes in such a way that a movie needs to be in both datasets before it ends up in the final merged dataframe.

In [ ]:
merge=pd.merge(movie_data,rating_data,on='imdb_title_id')
merge

,imdb_title_id,title,year,genre,duration,country,language,avg_vote,votes,total_votes,weighted_average_vote,mean_vote,us_voters_rating,non_us_voters_rating
0,tt0000009,Miss Jerry,1894,Romance,45,USA,NaN,5.9,154,154,5.9,5.9,6.4,6.0
1,tt0000574,The Story of the Kelly Gang,1906,"Biography, Crime, Drama",70,Australia,NaN,6.1,589,589,6.1,6.3,6.0,6.2
2,tt0001892,Den sorte drøm,1911,Drama,53,"Germany, Denmark",NaN,5.8,188,188,5.8,6.0,6.2,5.9
3,tt0002101,Cleopatra,1912,"Drama, History",100,USA,English,5.2,446,446,5.2,5.3,5.5,4.7
4,tt0002130,L'Inferno,1911,"Adventure, Drama, Fantasy",68,Italy,Italian,7.0,2237,2237,7.0,6.9,7.0,7.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
85850,tt9908390,Le lion,2020,Comedy,95,"France, Belgium",French,5.3,398,398,5.3,5.5,6.3,5.3
85851,tt9911196,De Beentjes van Sint-Hildegard,2020,"Comedy, Drama",103,Netherlands,"German, Dutch",7.7,724,724,7.7,7.9,6.8,7.7
85852,tt9911774,Padmavyuhathile Abhimanyu,2019,Drama,130,India,Malayalam,7.9,265,265,7.9,7.8,NaN,2.0
85853,tt9914286,Sokagin Çocuklari,2019,"Drama, Family",98,Turkey,Turkish,6.4,194,194,6.4,9.4,1.7,5.8


### 8.1.6 Genre Analysis

Movies often have multiple genres. Let's split them to see which genres are most common and how they relate to ratings. This is like unpacking a mixed bag of groceries to see what's inside.

Use `str.split()` and `explode()` to analyze genres.

In [133]:
merge['genre']=merge['genre'].str.split(',')
merge = merge.explode('genre')
merge

,imdb_title_id,title,year,genre,duration,country,language,avg_vote,votes,total_votes,weighted_average_vote,mean_vote,us_voters_rating,non_us_voters_rating
0,tt0000009,Miss Jerry,1894,Romance,45,USA,NaN,5.9,154,154,5.9,5.9,6.4,6.0
1,tt0000574,The Story of the Kelly Gang,1906,Biography,70,Australia,NaN,6.1,589,589,6.1,6.3,6.0,6.2
1,tt0000574,The Story of the Kelly Gang,1906,Crime,70,Australia,NaN,6.1,589,589,6.1,6.3,6.0,6.2
1,tt0000574,The Story of the Kelly Gang,1906,Drama,70,Australia,NaN,6.1,589,589,6.1,6.3,6.0,6.2
2,tt0001892,Den sorte drøm,1911,Drama,53,"Germany, Denmark",NaN,5.8,188,188,5.8,6.0,6.2,5.9
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
85851,tt9911196,De Beentjes van Sint-Hildegard,2020,Drama,103,Netherlands,"German, Dutch",7.7,724,724,7.7,7.9,6.8,7.7
85852,tt9911774,Padmavyuhathile Abhimanyu,2019,Drama,130,India,Malayalam,7.9,265,265,7.9,7.8,NaN,2.0
85853,tt9914286,Sokagin Çocuklari,2019,Drama,98,Turkey,Turkish,6.4,194,194,6.4,9.4,1.7,5.8
85853,tt9914286,Sokagin Çocuklari,2019,Family,98,Turkey,Turkish,6.4,194,194,6.4,9.4,1.7,5.8


### 8.1.5 new variable

One of the most popular genre combinations is romantic comedies like Notting Hill, Love Actually and Pulp Fiction. 
Make a new variable which is true if there is 'romance' and 'comedy' in the genre. How many romantic comedies are there in the data?

In [134]:
a=(merge['genre']==('Romance')).sum()
b=(merge['genre']==('Comedy')).sum()
a+b


np.int64(25064)

In [135]:
merge["Romance"] = merge['genre'] =='Romance'
merge["Comedy"] = merge['genre'] =='Comedy'
merge

,imdb_title_id,title,year,genre,duration,country,language,avg_vote,votes,total_votes,weighted_average_vote,mean_vote,us_voters_rating,non_us_voters_rating,Romance,Comedy
0,tt0000009,Miss Jerry,1894,Romance,45,USA,NaN,5.9,154,154,5.9,5.9,6.4,6.0,True,False
1,tt0000574,The Story of the Kelly Gang,1906,Biography,70,Australia,NaN,6.1,589,589,6.1,6.3,6.0,6.2,False,False
1,tt0000574,The Story of the Kelly Gang,1906,Crime,70,Australia,NaN,6.1,589,589,6.1,6.3,6.0,6.2,False,False
1,tt0000574,The Story of the Kelly Gang,1906,Drama,70,Australia,NaN,6.1,589,589,6.1,6.3,6.0,6.2,False,False
2,tt0001892,Den sorte drøm,1911,Drama,53,"Germany, Denmark",NaN,5.8,188,188,5.8,6.0,6.2,5.9,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
85851,tt9911196,De Beentjes van Sint-Hildegard,2020,Drama,103,Netherlands,"German, Dutch",7.7,724,724,7.7,7.9,6.8,7.7,False,False
85852,tt9911774,Padmavyuhathile Abhimanyu,2019,Drama,130,India,Malayalam,7.9,265,265,7.9,7.8,NaN,2.0,False,False
85853,tt9914286,Sokagin Çocuklari,2019,Drama,98,Turkey,Turkish,6.4,194,194,6.4,9.4,1.7,5.8,False,False
85853,tt9914286,Sokagin Çocuklari,2019,Family,98,Turkey,Turkish,6.4,194,194,6.4,9.4,1.7,5.8,False,False


In [136]:
RomanceComedy = merge.groupby('imdb_title_id')[['Romance','Comedy']].max()
(RomanceComedy["Romance"] & RomanceComedy["Comedy"]).sum()

np.int64(0)

### 8.1.8 Holywood movies


Let's focus on the 'Hollywood' movie scene, so select all movies from the USA.

Answer the following questions

1. What's the year of the oldest movie in the Hollywood data
2. What's the year of the youngest movie in this data

If there is only one oldest or youngest movie, give the name. If there are multiple, give the amount of movies.

In [139]:
merge.head()

,imdb_title_id,title,year,genre,duration,country,language,avg_vote,votes,total_votes,weighted_average_vote,mean_vote,us_voters_rating,non_us_voters_rating,Romance,Comedy
0,tt0000009,Miss Jerry,1894,Romance,45,USA,NaN,5.9,154,154,5.9,5.9,6.4,6.0,True,False
1,tt0000574,The Story of the Kelly Gang,1906,Biography,70,Australia,NaN,6.1,589,589,6.1,6.3,6.0,6.2,False,False
1,tt0000574,The Story of the Kelly Gang,1906,Crime,70,Australia,NaN,6.1,589,589,6.1,6.3,6.0,6.2,False,False
1,tt0000574,The Story of the Kelly Gang,1906,Drama,70,Australia,NaN,6.1,589,589,6.1,6.3,6.0,6.2,False,False
2,tt0001892,Den sorte drøm,1911,Drama,53,"Germany, Denmark",NaN,5.8,188,188,5.8,6.0,6.2,5.9,False,False


In [145]:
(merge['country']=='USA').sum()

np.int64(61745)

In [ ]:
merge[(merge['country']=="USA") & (merge['year']==merge['year'].min())] #oldest movie
#miss jerry

,imdb_title_id,title,year,genre,duration,country,language,avg_vote,votes,total_votes,weighted_average_vote,mean_vote,us_voters_rating,non_us_voters_rating,Romance,Comedy
0,tt0000009,Miss Jerry,1894,Romance,45,USA,NaN,5.9,154,154,5.9,5.9,6.4,6.0,True,False


In [157]:
merge[(merge['country']=="USA") & (merge['year']==merge['year'].max())] #youngest

,imdb_title_id,title,year,genre,duration,country,language,avg_vote,votes,total_votes,weighted_average_vote,mean_vote,us_voters_rating,non_us_voters_rating,Romance,Comedy
52039,tt0983946,Fantasy Island,2020,Action,109,USA,English,4.9,26823,26823,4.9,5.0,4.9,4.8,False,False
52039,tt0983946,Fantasy Island,2020,Adventure,109,USA,English,4.9,26823,26823,4.9,5.0,4.9,4.8,False,False
52039,tt0983946,Fantasy Island,2020,Fantasy,109,USA,English,4.9,26823,26823,4.9,5.0,4.9,4.8,False,False
52245,tt10003008,The Rental,2020,Horror,88,USA,English,5.7,9453,9453,5.7,5.7,6.0,5.7,False,False
52245,tt10003008,The Rental,2020,Thriller,88,USA,English,5.7,9453,9453,5.7,5.7,6.0,5.7,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
85774,tt9815998,Terapia di letto,2020,Drama,104,USA,English,5.2,807,807,5.2,5.3,5.3,5.0,False,False
85830,tt9887580,Bulletproof 2,2020,Action,97,USA,English,3.5,326,326,3.5,4.0,3.8,3.0,False,False
85830,tt9887580,Bulletproof 2,2020,Comedy,97,USA,English,3.5,326,326,3.5,4.0,3.8,3.0,False,False
85839,tt9898858,Coffee & Kareem,2020,Action,88,USA,English,5.1,10627,10627,5.1,5.1,5.1,5.0,False,False


In [154]:
a=merge[(merge['country']=="USA") & (merge['year']==merge['year'].max())] #youngest
a['title'].count()



np.int64(551)

### 8.1.9 Standardize (Z-score) the movie scores 'avg_vote'

Use the formula and not some python pakage for this. i.e. you can use functions such as mean(), median(), min(), max, std(), IQR(), etc.

Save this data in a new variable named 'standardized_vote'

In [159]:
standardize_vote = (merge['avg_vote']-merge['avg_vote'].mean())/merge['avg_vote'].std()   #(column-colum.mean/column.std)
standardize_vote

0       -0.019323
1        0.143970
1        0.143970
1        0.143970
2       -0.100970
           ...   
85851    1.450316
85852    1.613609
85853    0.388910
85853    0.388910
85854    0.633850
Name: avg_vote, Length: 175867, dtype: float64

## 8.2 Calculations 

In this section we are going to calculate some aggregate results per year and per decade. Specifically, we are interested in the average and standard deviation of the scores per year/decade. In this way we can hopefully answer the question which year/decade was rated the best for cinema. We are going to do this by writing a function that does these calculations for us

### 8.2.1 Write a function that calculates metrics per year

- Mean of 'avg_vote'
- Standard deviation of 'avg_vote'
- Mean of 'standardized_vote'
- Standard deviation of 'standardized_vote'
- Mean of 'duration'
- Standard deviation of 'duration'
- Total amount of movies

Use this function to answer the following questions:

1. In what year where most movies produced?
2. Which year had the highest average vote?
3. In what year were the standardized votes the least spread out?
4. Which year has the shortest movies?

### 8.2.2 Adapt your function to calculates metrics per decade

Only consider decades up to the 2010's.

Keep the same variables as before. Use the new function to answer the same questions as 8.2.1. Do your answers change?

## 8.3 Visualising results

### 8.3.1 Visualise the average scores (+/- SD) over time

Visualise the average scores per year. Also show the spread of each year's scores by showing the standard deviation.

There is a high chance your data will look weird, if so: fix it.

*Hint: remember we are working with a linear passage of time...*

### 8.3.2 Visualise the average scores (+/- SD) over time, and add the decade based data as an errorbar visual.

*Hint: this is a new element to visualize your data, so do some research!*

### 8.3.3 Change the previous plot but plot Us vs non-US votes.

You no longer need to show the standard deviation or decade errorplots.

Adjust your function to include this information.

### 8.3.4 Calculate the difference in rating between us votes and non-us votes

Create a function to calulate the difference in rating for each movie. Show this yearly difference in a histogram. Show the median and standard deviation.

### 8.3.5 Show the difference in rating between us votes and non-us votes on a plot

Show the caluculated difference on a plot. Also show the standard deviation per year.

### 8.3.6 Correlations and Relationships

Let's explore relationships between variables. For example, do longer movies get higher ratings? This is like checking if bigger portions lead to happier customers.

Use `corr()` for correlation matrix and visualize with heatmap and scatter plot.

### 8.4 Machine Learning Teaser: Predicting Movie Success

In your next course, Data Science Fundamentals, you'll learn about machine learning models. Why did we spend time cleaning and exploring data? Because good data leads to good predictions!

For example, based on our correlations, we could predict a movie's rating from its duration and genre. Here's a simple teaser using basic statistics (not ML yet):

- High correlation between US and non-US ratings suggests consistent quality perception.
- Longer movies might have slightly higher ratings, but not strongly.

In ML, we'd use these insights to choose models like linear regression for predictions. Excited for more? 🚀